resources/assets/lib/beatmap-discussions/editor-insertion-menu.tsx

```diff
@@ -0,0 +1,295 @@
+import { Portal } from 'portal';
+import * as React from 'react';
+import { Editor as SlateEditor, Element as SlateElement, Node as SlateNode, Point, Transforms } from 'slate';
+import { ReactEditor } from 'slate-react';
+import { SlateContext } from './slate-context';
+
+const editorClass = 'beatmap-discussion-editor';
+
+interface Props {
+  currentBeatmap: BeatmapJsonExtended;
+}
+
+export class EditorInsertionMenu extends React.Component<Props> {
+  static contextType = SlateContext;
+  bn = 'beatmap-discussion-editor-insertion-menu';
+  hideInsertMenuTimer?: number;
+  hoveredBlock: HTMLElement | undefined;
+  insertRef: React.RefObject<HTMLDivElement>;
+  menuPos: string | undefined;
+  mouseOver = false;
+  scrollContainer: HTMLElement | undefined;
+  // setTimeout delay is to prevent flashing when hovering the menu (the portal is not inside the container, so it fires a mouseleave)
+  throttledContainerMouseExit = _.throttle(() => { setTimeout(this.hideMenu.bind(this), 100); }, 10);
+  throttledContainerMouseMove = _.throttle(this.containerMouseMove.bind(this), 10);
+  throttledMenuMouseEnter = _.throttle(this.menuMouseEnter.bind(this), 10);
+  throttledMenuMouseExit = _.throttle(this.menuMouseLeave.bind(this), 10);
+  throttledScroll = _.throttle(this.forceHideMenu.bind(this), 10);
+
+  constructor(props: Props) {
+    super(props);
+
+    this.insertRef = React.createRef();
+  }
+
+  componentDidMount() {
+    if (this.insertRef.current) {
+      $(this.insertRef.current).on('mouseenter.editor-insert-menu', this.throttledMenuMouseEnter);
+      $(this.insertRef.current).on('mouseleave.editor-insert-menu', this.throttledMenuMouseExit);
+    }
+    $(window).on('scroll.editor-insert-menu', this.throttledScroll);
+  }
+
+  // updates cascade from our parent (slate editor), i.e. `componentDidUpdate` gets called on editor changes (typing/selection changes/etc)
+  componentDidUpdate() {
+    this.forceHideMenu();
+  }
+
+  componentWillUnmount() {
+    $(window).off('.editor-insert-menu');
+    if (this.scrollContainer) {
+      $(this.scrollContainer).off('.editor-insert-menu');
+    }
+    if (this.insertRef.current) {
+      $(this.insertRef.current).off('.editor-insert-menu');
+    }
+  }
+
+  containerMouseMove(event: JQuery.MouseMoveEvent) {
+    const block = event.target.closest(`.${editorClass}__block`);
+
+    if (
+      !block ||
+      !this.insertRef.current ||
+      !event.originalEvent ||
+      event.originalEvent.buttons > 0 // don't show while dragging
+    ) {
+      this.hoveredBlock = undefined;
+      return;
+    }
+
+    this.hoveredBlock = block;
+    const blockRect = block.getBoundingClientRect();
+    const cursorPos = {
+      x: event.originalEvent.clientX,
+      y: event.originalEvent.clientY,
+    };
+
+    // If we're past the half-way point of the block's height then put the menu below the block, otherwise put it above
+    if ((cursorPos?.y - blockRect.top) > (blockRect.height / 2)) {
```

Comment:
I wonder if `cursorPos?.y > (blockRect.top + blockRect.height / 2)` is easier to read